### DISCUSSION TITLE
Configuring dependencies at runtime, eg tokens for URLRoutingClient

### DISCUSSION THREAD

**Discussion by kamcma** [pointfreeco/swift-composable-architecture#1220] [Q&A]
`URLRoutingClient` from `swift-url-routing` is quite testable with its `override(_:with:)` methods. This makes it very tempting to use directly in a TCA environment, particularly in the new concurrency-beta branch, without the overhead of wrapping. My complication comes when I'd like to integrate OAuth token refreshing functionality.

In particular, constructing a live implementation like this would be very cool:
```swift
import AppAuth
import URLRouting

extension URLRoutingClient {
  static func appAuthLive<R: ParserPrinter>(
    router: R,
    session: URLSession = .shared,
    decoder: JSONDecoder = .init(),
    authState: OIDAuthState
  ) -> Self where R.Input == URLRequestData, R.Output == Route {
    return Self(
      request: { route in
        let (token, _): (String, String) = try await withCheckedThrowingContinuation { continuation in
          authState.performAction { (token: String?, id: String?, error: Error?) in
            guard
              let token = token,
              let id = id
            else {
              continuation.resume(throwing: error ?? OIDErrorUtilities.error(with: .tokenRefreshError, underlyingError: error, description: nil))
              return
            }
            continuation.resume(returning: (token, id))
          }
        }

        let request = try router
          .baseRequestData(.init(headers: ["Authorization": ["Bearer \(token)"]]))
          .request(for: route)

        return try await session.data(for: request)
      },
      decoder: decoder
    )
  }
}
```
If you're not familiar with the AppAuth library, the `OIDAuthState` object holds your OAuth token. Its `performAction()` method will either give you your token, or refresh it if necessary. I believe it has some internal synchronization logic to not have more than one refresh request in flight, no matter how many times you call `.performAction()` in quick succession.

My question isn't about AppAuth per se; you could use any OAuth logic or library of your choice. My question is: `OIDAuthState` won't exist on first launch, or after a logout, so I don't think I can statically configure `URLRoutingClient` like this. I understand from discussions elsewhere, eg #721, the environment is supposed to be a static set of dependencies. Does that foreclose abstracting away token logic like this? Ideally, I'd like it to be an implementation detail of the client dependency; I'm not really interested in it vis-à-vis my business logic, at least down in the child features downstream of being logged in.

I think I'm approaching this slightly wrong, and would be glad if anyone had any clarifying pointers, or maybe an alternate approach to achieve a similar end. Thanks!

**Comment by nayooti** [comment]
Maybe this answer helps.

https://github.com/pointfreeco/swift-composable-architecture/discussions/759
